ror-tools-core/src/main/scala/tech/beshu/ror/tools/core/patches/base/EsPatchExecutor.scala

```diff
@@ -0,0 +1,203 @@
+      case Some(metadata) if metadata.rorVersion == currentRorVersion && metadata.esVersion.major == currentEsVersion.major =>
+        validatePatchedFiles(metadata.patchedFilesMetadata) match {
+          case Right(()) =>
+            PatchedWithCurrentRorVersion(currentRorVersion)
+          case Left(invalidFiles) =>
+            IllegalFileModificationsDetectedInPatchedFiles(invalidFiles)
+        }
+      case Some(metadata) if metadata.rorVersion == currentRorVersion && !(metadata.esVersion.major == currentEsVersion.major) =>
+        PatchPerformedOnOtherEsVersion(currentEsVersion.render, metadata.esVersion.render)
+      case Some(metadata) =>
+        PatchedWithOtherRorVersion(currentRorVersion, metadata.rorVersion)
+      case None =>
+        // It is the situation, when the metadata file does not exist.
+        // In that case we perform additional checks, in order to detect corrupted patch:
+        // - we check whether the backup folder exists
+        // - we search for any patched jar files
+        val backupFolderExists = rorPluginDirectory.doesBackupFolderExist
+        val patchedJarFiles = searchForPatchedJarFiles() match {
+          case Left(files) => Some(files)
+          case Right(()) => None
+        }
+        if (backupFolderExists || patchedJarFiles.nonEmpty) {
+          CorruptedPatchWithoutValidMetadata(
+            backupFolderIsPresent = backupFolderExists,
+            patchedJarFiles = patchedJarFiles.getOrElse(List.empty)
+          )
+        }
+        else NotPatched
+    }
+  }
+
+  private def validatePatchedFiles(patchedFilesMetadata: List[FilePatchMetadata]): Either[List[os.Path], Unit] = {
+    patchedFilesMetadata.map { filePatchMetadata =>
+      val currentHash = FileUtils.calculateFileHash(filePatchMetadata.path.wrapped)
+      if (filePatchMetadata.hash == currentHash) Right(()) else Left(filePatchMetadata.path)
+    }.partitionMap(identity) match {
+      case (paths, _) if paths.nonEmpty => Left(paths)
+      case _ => Right(())
+    }
+  }
+
+  private def searchForPatchedJarFiles(): Either[List[PatchedJarFile], Unit] = {
+    JarManifestModifier.findPatchedFiles(rorPluginDirectory.esDirectory) match {
+      case Nil => Right(())
+      case nel => Left(nel)
+    }
+  }
+
+}
+
+object EsPatchExecutor {
+
+  sealed trait EsPatchStatus
+
+  object EsPatchStatus {
+    final case class PatchedWithCurrentRorVersion(currentRorVersion: String) extends EsPatchStatus
+
+    final case class PatchPerformedOnOtherEsVersion(currentEsVersion: String, patchPerformedOnEsVersion: String) extends EsPatchStatus
+
+    final case class PatchedWithOtherRorVersion(currentRorVersion: String, patchedByRorVersion: String) extends EsPatchStatus
+
+    final case class CorruptedPatchWithoutValidMetadata(backupFolderIsPresent: Boolean, patchedJarFiles: List[PatchedJarFile]) extends EsPatchStatus
+
+    final case class IllegalFileModificationsDetectedInPatchedFiles(files: List[os.Path]) extends EsPatchStatus
```

Comment:
CorruptedPatchWithoutValidMetadata and IllegalFileModificationsDetectedInPatchedFiles are similar cases, don't they?

They refer to a corrupted ES installation, but the reasons are different. Maybe we should have another subhierarchy, or at least the names should be similar.